src/uhttp.c

```diff
@@ -1053,70 +1059,83 @@ HTTP_CLIENT_RESULT uhttp_client_open(HTTP_CLIENT_HANDLE handle, const char* host
+            }
 
-            if (http_data->x509_cert != NULL && http_data->x509_pk != NULL)
+            if (mallocAndStrcpy_s(&http_data->host_name, host) != 0)
             {
-                if (xio_setoption(http_data->xio_handle, SU_OPTION_X509_CERT, http_data->x509_cert) != 0 || xio_setoption(http_data->xio_handle, SU_OPTION_X509_PRIVATE_KEY, http_data->x509_pk) != 0)
-                {
-                    LogError("Failed setting x509 certificate");
-                    result = HTTP_CLIENT_ERROR;
-                    free(http_data->host_name);
-                    http_data->host_name = NULL;
-                    http_data->on_connect = NULL;
-                    http_data->connect_user_ctx = NULL;
-                    http_data->port_num = 0;
-                }
+                LogError("copying hostname has failed");
+                result = HTTP_CLIENT_ERROR;
             }
-            if (result == HTTP_CLIENT_OK && http_data->certificate != NULL)
+            /* Codes_SRS_UHTTP_07_007: [http_client_connect shall attempt to open the xio_handle. ] */
+            else
             {
-                if (xio_setoption(http_data->xio_handle, OPTION_TRUSTED_CERT, http_data->certificate) != 0)
+                result = HTTP_CLIENT_OK;
+                http_data->recv_msg.recv_state = state_opening;
+                http_data->on_connect = on_connect;
+                http_data->connect_user_ctx = callback_ctx;
+                http_data->port_num = port_num;
+
+                if ((http_data->x509_cert != NULL) && (http_data->x509_pk != NULL))
                 {
-                    LogError("Failed setting Trusted certificate");
-                    result = HTTP_CLIENT_ERROR;
-                    free(http_data->host_name);
-                    http_data->host_name = NULL;
-                    http_data->on_connect = NULL;
-                    http_data->connect_user_ctx = NULL;
-                    http_data->port_num = 0;
+                    if ((xio_setoption(http_data->xio_handle, SU_OPTION_X509_CERT, http_data->x509_cert) != 0) ||
+                        (xio_setoption(http_data->xio_handle, SU_OPTION_X509_PRIVATE_KEY, http_data->x509_pk) != 0))
+                    {
+                        LogError("Failed setting x509 certificate");
+                        result = HTTP_CLIENT_ERROR;
+                        free(http_data->host_name);
+                        http_data->host_name = NULL;
+                        http_data->on_connect = NULL;
+                        http_data->connect_user_ctx = NULL;
+                        http_data->port_num = 0;
+                    }
                 }
-            }
 
-            if (result == HTTP_CLIENT_OK)
-            {
-                if (xio_open(http_data->xio_handle, on_xio_open_complete, http_data, on_bytes_received, http_data, on_io_error, http_data) != 0)
+                if ((result == HTTP_CLIENT_OK) && (http_data->certificate != NULL))
```

Comment:
Why the certificate must be not null?
Will this prevent the usage of the module in systems where xio can get the certificates out of the cert store?
This is a significant change.

Reply:
I didn't change this line except for alignment - another if check was added to address your comments - and @danewalton 's comments to add the parentheses. 

I don't understand your comment here though: the cert must not be null to be set as an option. This will not fail open.
I wanted to be sure so I've checked `tlsio_schannel.c:1444` - if the option is NULL, the default stores are to be used. If not-null,  `SCH_CRED_MANUAL_CRED_VALIDATION` is performed. For schannel users it's actually advised to have this option to the default NULL.

Reply:
This comment was not good. Sorry. But thanks for checking anyway, that's good!

---
In reply to: [339241331](https://github.com/Azure/azure-uhttp-c/pull/49#discussion_r339241331) [](ancestors = 339241331)